src/Simulation/Simulators/CommonNativeSimulator/QubitManager.cs

```diff
@@ -8,46 +8,48 @@
 
 namespace Microsoft.Quantum.Simulation.Simulators
 {
-    public partial class QuantumSimulator
+    public partial class CommonNativeSimulator
     {
-        class QSimQubitManager : QubitManager
+        protected class QSimQubitManager : QubitManager
         {
             readonly bool throwOnReleasingQubitsNotInZeroState;
 
-            public uint SimulatorId { get; private set; }
+            public CommonNativeSimulator? Simulator { get; set; }   // Must not be nullable (and public). But we cannot 
+                // initialize it properly _during construction_. We initialize it _after construction_. 
+                // That is why it is nullable and public.
 
             public QSimQubitManager(bool throwOnReleasingQubitsNotInZeroState = true, long qubitCapacity = 32, bool mayExtendCapacity = true, bool disableBorrowing = false)
                 : base(qubitCapacity, mayExtendCapacity, disableBorrowing)
             {
                 this.throwOnReleasingQubitsNotInZeroState = throwOnReleasingQubitsNotInZeroState;
             }
 
-            public void Init(uint simulatorId)
-            {
-                this.SimulatorId = simulatorId;
-            }
-
             /// <summary>
             ///  The max number used as qubit id so far.
             /// </summary>
             public long MaxId { get; private set; }
 
             public override Qubit CreateQubitObject(long id)
             {
-                Debug.Assert(id < 50, "Using a qubit id > 50. This is a full-state simulator! Validating ids uniquenes might start becoming slow.");
+                Debug.Assert(id < 50, "Using a qubit id > 50. This is a full-state simulator! Validating ids uniqueness might start becoming slow.");
```

Comment:
Sparse simulator may handle more than 50 qubits.

Reply:
So what should be the number here instead of 50?

Reply:
Theoretically there's no artificial limit. You may have millions of qubits as long as they are all in zero state. However, it's hard to imagine a meaningful program that would have that many in the zero state. I believe there're tests that check functionality of the simulator with a few thousand qubits. But the point is - the check for full state simulator and sparse simulator should be substantially different.